### ISSUE TITLE
mutation model

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/natural-selection#98]
In https://github.com/phetsims/natural-selection/issues/97#issuecomment-637875545, I described how mutation is currently modeled:

> When a bunny is mutated, it is made homozygous - both of its alleles are set to the mutant allele. This guarantees that the mutation will be reflected in the bunny's appearance (phenotype). In reality, a mutation is a change in one of the inherited genes, not both. I discussed this with @amanda-phet awhile back, and she verified that making the mutant bunny homozygous is the desired behavior.
> 
> So for example, suppose that brown fur is applied as a recessive mutation. The only fur allele in the gene pool would have been white fur, so the bunny will inherit 2 white fur alleles from it parents. Then the mutation occurs. If we mutated only 1 of the inherited alleles, then the mutant bunny would be white, because white fur is dominant and it would still have 1 white fur allele. By mutating both alleles (making the bunny homozygous) whether the mutant allele is dominant or recessive is irrelevant, and the bunny's appearance will correspond to the mutant allele.
> 
> And to clarify... This is for when a mutation first occurs. After mutation has occurred, the mutant allele is just another gene in the gene pool, and subsequent matings use the childAlleles approach described above to determine the child's alleles.

I think we should reconsider whether mutants should always be homozygous. When I encountered this implementation in the Java model, it seemed problematic to me. The science is incorrect, and it makes it impossible to learn the difference between a dominant and recessive mutation.

Suppose that I press the Add Mutation button for "brown fur, recessive".  With  homozygous mutants, the mutant will always have brown fur, and the resulting pedigree will always be this:

<img width="591" alt="screenshot_373" src="https://user-images.githubusercontent.com/3046552/83778732-56108200-a648-11ea-8b5b-e96fdba10e43.png">

But if the mutation was recessive, why does it always appear immediately in the phenotype? If a mutation were recessive, this is a correct outcome, which our model cannot produce:

<img width="591" alt="screenshot_374" src="https://user-images.githubusercontent.com/3046552/83778949-92dc7900-a648-11ea-8a99-e4d96ba8f46d.png">

The mutant in the above pedigree has genotype 'Ff', and because 'brown fur' is recessive ('f') the bunny is white.   Brown fur will eventually appear in the population, but not until an individual is born whose genotype is 'ff' (homozygous recessive), as in this example:

<img width="590" alt="screenshot_375" src="https://user-images.githubusercontent.com/3046552/83779233-eb137b00-a648-11ea-9178-06e480c496da.png">

There are also issues when choosing "brown fur, dominant".  It's appropriate that the mutant bunny is brown, because brown fur is dominant.  But looking at the Pedigree graph, why is the resulting bunny 'FF' (homozygous dominant) and not 'Ff' or 'fF' (heterozygous)?  How/why did a mutation affect _both_ inherited alleles? 

E.g., why this:

<img width="591" alt="screenshot_376" src="https://user-images.githubusercontent.com/3046552/83786255-b35c0180-a64f-11ea-8623-759af086cbba.png">

... and not this?

<img width="594" alt="screenshot_377" src="https://user-images.githubusercontent.com/3046552/83786283-bb1ba600-a64f-11ea-9402-f1c13e1ce229.png">

So I want to verify that we are deliberately (a) choosing incorrect science, and (b) forgoing the above scenarios, which make the important point that a mutation does not affect both inherited alleles, and (if recessive) will not immediately result in a change in appearance. A recessive mutation will not appear in the population until generations after the mutation occurred.

@amanda-phet @kathy-phet @ariel-phet thoughts?

**Comment by pixelzoom** [comment]
In the above commit, I added a query parameter that selects between 2 mutation models:

`?homozygousMutants=true` (default):  When a mutation is applied, the mutation affects both of its inherited alleles, and the resulting bunny is homozygous. Examples: 
* If the mutation if "brown fur, dominant", then the bunny is 'FF'.  
* If the mutation is "brown fur, recessive", then the bunny is 'ff'.

`?homozygousMutants=false`:  When a mutation is applied, the mutation randomly affects one of its inherited alleles, and the resulting bunny is heterozygous. Examples: 
* If the mutation if "brown fur, dominant", and the mutation changes the father allele, then the bunny is 'Ff'. (Genotype abbreviation 'Ff' reads 'father allele is brown fur, mother allele is white fur.') 
* If the mutation if "brown fur, dominant", and the mutation changes the mother allele, then the bunny is 'fF'.
* If the mutation is "brown fur, recessive", and the mutation changes the father allele, then the bunny is 'fF'.
* If the mutation is "brown fur, recessive", and the mutation changes the mother allele, then the bunny is 'Ff'.

You can experiment with this query parameter in master, or in [1.0.0-dev.48](https://phet-dev.colorado.edu/html/natural-selection/1.0.0-dev.48/phet/natural-selection_en_phet.html).

**Comment by amanda-phet** [comment]
Augus 1, 2019 we had a discussion about this model simplification. While Megan agreed that a newly-mutated rabbit would in reality be heterozygous, we all agreed that the model simplification was a reasonable one so we don't have to wait as many generations for the gene to be expressed. It seems like this is resurfacing because we are being explicit about genotypes in the Pedigree, which could cause some confusion, so I'm fine with us revisiting this decision. 

I think if we make the mutation heterozygous, it will take too many generations for the gene to be expressed, and the bunnies will probably take over the world (or die) before we have a chance to see natural selection take place. However I will play with the query parameters you provided (likely in conjunction with the `PercentToKill` parameters, so we have reasonable rates of bunnies dying).

**Comment by pixelzoom** [comment]
Sorry, I wasn't involved in the 8/1/2019 discussion.  And yes, I'm raising the issue because it's explicitly shown in genotype abbreviation that appears in the Pedigree graph.
 
Note that it's also possible to randomly create some mutations homozygous and some heterozygous, if that accomplished the goals.  And I could easily modify the query parameter to something like `mutate=` with values 'homozygous', 'heterozygous', and 'random', if that would help.

**Comment by pixelzoom** [comment]
Sorry, I wasn't involved in the 8/1/2019 discussion.  Yes, I'm raising the issue because it's explicitly shown in the genotype abbreviation that appears in the Pedigree graph.  But I also realized that immediately expressing a _recessive_ mutation is pedantically incorrect.
 
Note that it's also possible to randomly create some mutations homozygous and some heterozygous, if that accomplishes the goals.  And I could easily modify the query parameter to something like `mutate=` with values 'homozygous', 'heterozygous', and 'random', if that would help.

**Comment by kathy-phet** [comment]
I think having the query parameters to change the behavior is nice and would favor having those - for teachers' flexibility.

**Comment by pixelzoom** [comment]
Hmmm... I hadn't considered making the query parameter public-facing.  It's easy to do, but it's going to make "tuning" the model considerably more difficult, if there is specific behavior that you want to be easy to achieve (e.g. bunnies don't die too soon, bunnies don't take over the world too soon, easy to achieve a stable population).

**Comment by pixelzoom** [comment]
Hmmm... I hadn't considered making the query parameter public-facing.   It's easy to make it public. But it's going to make "tuning" the model considerably more difficult, if there is specific behavior that you want to be easy to achieve (e.g. bunnies don't die too soon, bunnies don't take over the world too soon, easy to achieve a stable population).  So with this flexibility will come great cost.

**Comment by amanda-phet** [comment]
OK, so here is what I'm noticing. I am running a version with `homozygousMutants=false` and a version without. They both have the same parameters for percentages to kill. 

Model 1: https://phet-dev.colorado.edu/html/natural-selection/1.0.0-dev.48/phet/natural-selection_en_phet.html?secondsPerGeneration=2&homozygousMutants=false&wolvesPercentToKill=.35,.4&wolvesEnvironmentMultiplier=2.3&toughFoodPercentToKill=.4,.7&shortTeethMultiplier=1.4

Model 2: https://phet-dev.colorado.edu/html/natural-selection/1.0.0-dev.48/phet/natural-selection_en_phet.html?secondsPerGeneration=2&wolvesPercentToKill=.35,.4&wolvesEnvironmentMultiplier=2.3&toughFoodPercentToKill=.4,.7&shortTeethMultiplier=1.4

## Experiment 1: 
Add a dominant fur mutation immediately, add wolves at the start of generation 3. 

### Outcomes:
#### Model 1: Brown bunnies take over at the start of generation 12. Good.
![image](https://user-images.githubusercontent.com/7852916/83909665-73b41900-a726-11ea-8bd5-b1ce0e5c10d1.png)

#### Model 2: Brown bunnies take over at the start of generation 10. Good.
![image](https://user-images.githubusercontent.com/7852916/83909746-92b2ab00-a726-11ea-8f54-d4773cadb757.png)


### Conclusion: 
For the dominant fur mutation, a heterozygous mutation accomplishes the learning goal of seeing the selection for white bunnies, and seeing the brown bunnies take over the world. 

## Experiment 2: 
Add a recessive fur mutation immediately, add wolves at the start of generation 5. 

### Outcomes: 
#### Model 1: Brown fur was not expressed until generation 4, and was only 1-2 bunnies (hard to tell in the proportions graph and I can't remember exactly). The bunnies all died during generation 10. Bad.
![image](https://user-images.githubusercontent.com/7852916/83909987-09e83f00-a727-11ea-9588-1e02edee1f12.png)

#### Model 2: Brown fur was expressed in generation 1 (1/6 bunnies). Brown bunnies are present in 100% of the population in generation 10. Brown bunnies take over the world at the start of generation 14. Good.

### Conclusion: 
For the recessive fur mutation, a heterozygous mutation does not accomplish the learning goal of seeing bunnies with an advantageous heritable trait increase in proportion to organisms lacking this this trait. It is very powerful for student to see that even a recessive trait can become "dominant" if it allows for better survival. 

----- 

My thoughts: I'd like this to work so that we don't need the model simplification (ie. homozygous alleles), but it takes too many generations for the gene to even present itself, and the white bunnies will take over before we even have a chance for the brown bunnies to mate enough. So perhaps that means we need to kill fewer bunnies and use a higher multiplier, and I can experiment with that, but I'm not sure it will happen.

**Comment by pixelzoom** [comment]
Here's another idea, thinking out loud.

For recessive mutations:
* make the original mutant(s) heterozygous
* change the mating model to guarantee that a recessive mutation appears in the phenotype as soon as there are 2 individuals with the mutant allele who can be mated

**Comment by pixelzoom** [comment]
Here's another idea, thinking out loud.

* For all mutations, make the original mutant(s) heterozygous.
* For recessive mutations, change the mating model to guarantee that a recessive mutation appears in the phenotype as soon as there are 2 individuals with the mutant allele who can be mated

**Comment by amanda-phet** [comment]
>Here's another idea, thinking out loud.

>* For all mutations, make the original mutant(s) heterozygous.
>* For recessive mutations, change the mating model to guarantee that a recessive mutation appears in the phenotype as soon as there are 2 individuals with the mutant allele who can be mated

I would love to consider this kind of direction!

**Comment by amanda-phet** [comment]
@pixelzoom and I discussed this 6/8/20 and decided to try changing the model so that as soon as there are at least 2 bunnies with the recessive mutation they will mate with each other. This means that within 2 generations, we will see the trait expressed.

**Comment by pixelzoom** [comment]
To add to what @amanda-phet said in https://github.com/phetsims/natural-selection/issues/98#issuecomment-641468178... We also decided to try what she described in addition to making mutants heterozygous instead of homozygous.

**Comment by ariel-phet** [comment]
I think @amanda-phet has done a nice job of focusing on the learning goal (getting users to see the change occur easily).

I would have a few pieces of input:
1. The Java sim is certainly quite popular and uses this homozygous mutation model. So if it is difficult to find a satisfactory tuning that accomplishes the goals sticking with this model (despite certain issues) might be reasonable.

2. If a satisfactory tuning can be found for heterozygous mutations we might want to consider some A/B testing interviews. Will students make the connections as quickly if things take a couple of generations? The generations do happen quickly now so maybe this would be OK

3. If (despite the complexity) we have a way for both models to work well and are considering public facing query parameters, I would actually advocate for this to be an explicit feature of the sim. I would think the sim would start with the homozygous model and then the user could switch to the more realistic heterozygous model. So you first get the idea, see the quick changes, etc. But maybe have some sort of more hidden "advanced" control (perhaps popping up a dialog) that allows you to change the model. But perhaps with the proper tuning and proofing out through interviews only a heterozygous model would be needed. 

4. And a very hair-brained thought -- if we really liked the heterozygous model but it wasn't quite getting the idea across quickly enough could we have a visual indicator of which bunnies had mutated (something like a halo, a dotted line around them or something). Some cue that something HAS changed, but not show that the phenotype has yet changed. Not well thought out, but just brainstorming.

**Comment by amanda-phet** [comment]
I like the mutation icon, but wouldn't want it to persist for too long... I guess it would only be there as long as that bunny is alive (could be many bunnies if the population is large). 

I look forward to seeing how this all works once we can get those bunnies with recessive traits mating faster.

**Comment by kathy-phet** [comment]
> I look forward to seeing how this all works once we can get those bunnies with recessive traits mating faster.

We can discuss this tomorrow. But if bunnies with recessive traits mate disproportionately faster, that is going to mess up the genetics inheritance probability and the statistics.

**Comment by pixelzoom** [comment]
@kathy-phet I'm not sure what you mean by "mate disproportionately faster" - all bunnies mate at the same rate, and the pairing is currently random.  What we were considering was to prefer pairing bunnies that have a recessive mutation with other bunnies that have the same recessive mutation, so that the recessive gene is expressed more quickly than if it were random.  That seems a lot less harmful than homogenous mutants, which totally throws the statistics off, and results in no difference between dominant and recessive mutations.

**Comment by kathy-phet** [comment]
Can you make that a query parameter initially so that we can look at the statistics of the distributions? It seems to me that something that is not "random" would cause a change in the expected outcome when you do controlled experiments by using the pre-set populations. 
@amanda-phet - Can we see if Megan is available to join a meeting with us soon?

**Comment by amanda-phet** [comment]
6/11/20 design meeting: we would like to move forward with this model simplification (mutations will be heterozygous, and original mutant bunnies will mate with another bunny with the recessive gene as soon as possible). 

We'd like to follow up with Megan about this simplification, since she had originally approved the homozygous mutation simplification. Also check with her about the model applying the mutation to 1/7 of the population, as well as adding a mutation indication on mutant bunnies.

**Comment by pixelzoom** [comment]
Also from 6/11/20 design meeting: @amanda-phet doesn't like the mutation icons in the environment view.  Some things to try: a different size, a different location on the bunnies (compatible with the red 'X'), a different color, a different icon, ...

**Comment by pixelzoom** [comment]
@kathy-phet had concerns about introducing a mutation into multiple individuals simultaneously. Wondering if that happens in reality, and if it is problematic for the sim.  Would like to discuss with Megan.

**Comment by pixelzoom** [comment]
n 6/11/2020 design meeting discussion of #98, I mentioned that mating 2 heterozygous recessive mutants could result in more than 1 offspring that has the mutation in the phenotype. The team corrected me, and said that only 1 offspring would have the mutation in the phenotype. Thinking about this more, I believe that I was correct.  

See https://github.com/phetsims/natural-selection/issues/97#issuecomment-643624507 for a table showing the possible outcomes of the inheritance model.   When mating 2 heterozygous parents (e.g. 'Ff' ) there are 2 possible outcomes for offspring: FF, Ff, fF, ff **or** FF, FF, ff, ff.  In the latter outcome, there are 2 offspring that have the mutation in their phenotype.  In both outcomes, parents donate their alleles with equal frequency (the requirement of Mendelian inheritance). But which outcome occurs depends on the random order in which the alleles are donated.

**Comment by amanda-phet** [comment]
6/15/20 meeting. 

We are interested in pursuing heterozygous mutations, and in the recessive case, breeding the mutant with another bunny with the recessive gene (a child) as soon as possible.

For immediate feedback, we are still interested in "labeling" the bunnies in the environment to show feedback that mutants exist in the population, but not in a way that is so distracting. 

Leaning toward removing badges from the environment, but still need to tune the model to be sure recessive traits can take over.

**Comment by amanda-phet** [comment]
More ideas for tuning the model: 
- might have to force some bunnies to mate beyond just the first one
- perhaps if there are 2 or fewer brown bunnies, wolves can't eat any brown bunnies

**Comment by pixelzoom** [comment]
6/15/20 design meeting more:

We decided that we won't do anything special related to recessive mutations that appear in an initial population that is set via query parameters.  There are no "original mutants" in that population.

**Comment by pixelzoom** [comment]
In the above commits, the mutation icon is no longer shown for bunnies in the environment. For development purposes (useful to me), they can be turned on using `?labelMutants`.

**Comment by pixelzoom** [comment]
@amanda-phet All of the requested changes have been made. You can review in master or [1.0.0-dev.53](https://phet-dev.colorado.edu/html/natural-selection/1.0.0-dev.53/phet/natural-selection_en_phet.html).

Summary:

* Mutants are not labeled in the environment. If you want to label them, use `?labelMutants` (internal use only).
* Mutants are heterozygous. If you want homozygous mutants, use `?homozygousMutants` (internal use only).
* Original mutants with a recessive mutation are mated as soon as possible ("mated eagerly"), pairing them with another bunny that has the same recessive mutant allele.
* For a mutation that was applied in generation N: If there is a sibling with the mutant allele, then the mating occurs in generation N+1.   If there is no sibling with the mutant allele, then the mating occurs with a child in generation N+2.  When mating occurs, 1 child will be homozygous recessive and will therefore show the mutation in its phenotype.
* After an original mutant has mated eagerly, all future mating is treated normally, by random pairing with another bunny.
* Recessive mutations that are applied via query parameters are not mated eagerly.

**Comment by amanda-phet** [comment]
I'm still having a hard time keeping the brown bunnies alive when the mutation is recessive. I'm trying to tweak `wolvesPercentToKill` and `wolvesEnvironmentMultiplier` as well. The best luck I've had is waiting to add the mutation so that more than one bunny gets it, but they still all die (it just takes longer). 

I'd like to consider not just the mutant bunny mating "eagerly" but any of the bunnies with the recessive gene. Happy to discuss this on Thursday, but if you have any other ideas we could try let me know.

**Comment by amanda-phet** [comment]
6/18/20 More ideas:
- A mutant bunny could have a 5th homozygote bunny  
- Have the selection factor ignore preferred phenotype if fewer than 4 exist
- Environmental factors don't apply to recessive bunnies until they have mated eagerly 

@pixelzoom will try all of these suggestions

**Comment by pixelzoom** [comment]
A few notes....

>  - [x] Selection factors don't apply to recessive bunnies until they have mated eagerly

This one was relatively easy.  Selection factors are ignored until a homozygous-recessive mutant has mated eagerly.

> - [x] A mutant bunny could have a 5th homozygote bunny

This one was complicated.  The genotype for the recessive mutation was easy, it always homozygous recessive. The genotype for the other 2 genes was more complicated.  For those other 2 genes, a random cell from the Punnett square containing a dominant allele will be preferred (since dominant alleles would be preferred in reality).  If no such cell exists, then any random cell is is chosen.  Note that this extra offspring will skew the distribution for all alleles, not just the mutation.

> - [ ] Have the selection factor ignore preferred phenotype if fewer than N exist

I'm not clear on what this means, so I need to discuss with @amanda-phet.

**Comment by kathy-phet** [comment]
Chris, 

The last one means that if there are fewer than say 3 brown bunnies, and the wolves come out to eat and it’s on brown ground, there is 0% chance they will eat brown bunnies. So it basically overwrites the general rule if there are few of the brown bunnies. Or whichever phenotype is preferred against the current selection factors for that generation. 

K

**Comment by pixelzoom** [comment]
@kathy-phet thanks for clarifying. So that last one is independent of whether the phenotype is the result of a dominant or recessive mutation.  That's what had me confused.

**Comment by pixelzoom** [comment]
> - [x] Have the selection factor ignore preferred phenotype if fewer than N exist

This requirement is now implemented in master, like this:

* If the population of bunnies with long teeth is <= `?minBunniesForFood` (default 5), then no bunnies with long teeth will starve to death.

* If the environment is the equator, and the population of brown bunnies is <= `?minBunniesForWolves` (default 5), then no brown bunnies will be eaten by wolves.

* If the environment is the arctic, and the population of white bunnies is <= `?minBunniesForWolves` (default 5), then no white bunnies will be eaten by wolves.

**Comment by pixelzoom** [comment]
All of the requested changes listed in https://github.com/phetsims/natural-selection/issues/98#issuecomment-646275437 are implemented. Test drive in master or [1.0.0-dev.63](https://phet-dev.colorado.edu/html/natural-selection/1.0.0-dev.63/phet/natural-selection_en_phet.html).  I'll review this with @amanda-phet to make sure that what I implemented is what she was expecting.

**Comment by kathy-phet** [comment]
Hi Chris, 

Shouldn't this one just also be "if" tough food selection factor is "on"? 
If the population of bunnies with long teeth is <= ?minBunniesForFood (default 5), then no bunnies with long teeth will starve to death.

Wolves could still eat long teeth bunnies that is. 

I'm curious how this will all behave when Amanda starts working on tuning it.

**Comment by pixelzoom** [comment]
> Shouldn't this one just also be "if" tough food selection factor is "on"? 
> If the population of bunnies with long teeth is <= ?minBunniesForFood (default 5), then no bunnies with long teeth will starve to death.

Correct.  I thought that goes without saying, because bunnies only starve if a food-selection factor is turned on. 

> Wolves could still eat long teeth bunnies that is.

Correct.  Environmental factors are independently applied.  Wolves select for fur color, and ignore other traits.  Food (starvation) selects for teeth, and ignore other traits.

**Comment by pixelzoom** [comment]
7/7/2020 @amanda-phet and I review this via Zoom, and this is looking good.  She'll experiment with the new query parameters (`minBunniesForFood`, `minBunniesForWolves`) and get back to me with change requests.

**Comment by amanda-phet** [comment]
I followed the same procedure as in https://github.com/phetsims/natural-selection/issues/98#issuecomment-639694444 and for experiment 2, with a recessive mutation, brown bunnies take over the world at the start of generation 16. This is using the same percentages (`wolvesPercentToKill=.35,.4&wolvesEnvironmentMultiplier=2.3`). 

If we want the bunnies to take over at generation 15 (just like with the dominant mutation) then I would use `minBunniesForWolves=6` but I think it's good at 5, myself.

**Comment by pixelzoom** [comment]
7/7/2020 design meeting @amanda-phet @kathy-phet @ariel-phet 

`minBunniesForWolves=6` makes bunnies take over earlier. But  `minBunniesForWolves=5` feels better, so leave as is.

@amanda-phet still needs to confirm the value for `minBunniesForFood`.

**Comment by amanda-phet** [comment]
See #86 

`minBunniesForWolves=5` and `minBunniesForFood=4` seem to work well.

**Comment by pixelzoom** [comment]
`minBunniesForWolves=5` and `minBunniesForFood=4` in the above commit.  I'm going to close this issue, and we'll handle any further tuning of those query parameters in #86.

**Comment by pixelzoom** [comment]
Reopening.  There's one bit of cleanup left to do here.  

@amanda-phet What shall we do about the `homozygousMutants` query parameter?  

In https://github.com/phetsims/natural-selection/issues/98#issuecomment-641772442, @ariel-phet proposed making this feature publicly visible, and perhaps even the default:

> If (despite the complexity) we have a way for both models to work well and are considering public facing query parameters, I would actually advocate for this to be an explicit feature of the sim. I would think the sim would start with the homozygous model and then the user could switch to the more realistic heterozygous model. So you first get the idea, see the quick changes, etc. But maybe have some sort of more hidden "advanced" control (perhaps popping up a dialog) that allows you to change the model. But perhaps with the proper tuning and proofing out through interviews only a heterozygous model would be needed.

My preference would be to delete this query parameter (and the associated code) altogether.  The science is incorrect, heterozygous mutants appear to be working nicely, and we don't need the added complexity of trying to tune the model to work for homozygous mutants.

**Comment by amanda-phet** [comment]
>My preference would be to delete this query parameter (and the associated code) altogether. The science is incorrect, heterozygous mutants appear to be working nicely, and we don't need the added complexity of trying to tune the model to work for homozygous mutants.

I 100% agree. The model is working really great right now, and I think maintaining that code is too much of a headache. I'm very happy with the mutation model!

**Comment by pixelzoom** [comment]
Reopening and assigning to me to remove `homozygousMutants`.   I'll defer this until tomorrow. @ariel-phet if you have strong feelings otherwise, let me know asap.

**Comment by pixelzoom** [comment]
`homozygousMutants` has been removed. Closing.

**Comment by pixelzoom** [comment]
@amanda-phet FYI, related to an overhaul of Food.js...

As I mentioned over in https://github.com/phetsims/natural-selection/issues/152#issuecomment-670234799...  `minBunniesForWolves=5` and `minBunniesForFood=4` are both "minimum population" constraints.  This sim has several of these constraints, and it's confusing/error-prone that some of them (like these 2) are specified as <=, while others are specified as <.   So I standardized on <, and these constraints are now < 6 and < 5 respectively. 

I've also eliminated the `minBunniesForWolves` and `minBunniesForFood` query parameters, and moved them to internal constants.  They are now `WolfCollection.MIN_BUNNIES` and `Food.TOUGH_FOOD_MIN_LONG_TEETH` respectively.
